### PR TITLE
Support for Xilinx Zynq UltraScale+ MPSoC

### PR DESCRIPTION
**Pull request by sorenb-xlnx**
This series provides the initial support for [Xilinx Zynq UltraScale+ MPSoC](http://www.xilinx.com/products/silicon-devices/soc/zynq-ultrascale-mpsoc.html). The ATF runs on the quad A53 cluster of the SoC.
In the current state, this provides enough to boot up the system and a shallow implementation of suspend.

### CODE REVIEW

**Comment by danh-arm** [comment]
Can one of the admins verify this patch?

**Comment by danh-arm** [comment]
Can one of the admins verify this patch?

File: plat/arm/common/aarch64/arm_common.c [hunk 1]
```diff
@@ -150,6 +151,7 @@ void arm_configure_sys_timer(void)
 
 	reg_val = (1 << CNTNSAR_NS_SHIFT(PLAT_ARM_NSTIMER_FRAME_ID));
 	mmio_write_32(ARM_SYS_TIMCTL_BASE + CNTNSAR, reg_val);
+#endif
```
Comment (danh-arm) [inline plat/arm/common/aarch64/arm_common.c]: I can't see anywhere in your other patches where you're using this function, so is this change still needed?

File: drivers/cadence/uart/cdns_console.S [hunk 2]
```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2013-2014, ARM Limited and Contributors. All rights reserved.
```
Comment (danh-arm) [inline drivers/cadence/uart/cdns_console.S]: minor: Should use 2016 throughout

File: include/drivers/cadence/cdns_uart.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2014, Xilinx Inc.
+ * Written by Edgar E. Iglesias.
```
Comment (danh-arm) [inline include/drivers/cadence/cdns_uart.h]: Please use the standard boilerplate text

File: drivers/cadence/uart/cdns_console.S [hunk 4]
```diff
@@ -0,0 +1,127 @@
+/*
```
Comment (danh-arm) [inline drivers/cadence/uart/cdns_console.S]: Does this uart have a part number or some other way to distinguish it? If so can this be reflected in the filenames? If this is only found in the Xilinx SoCs then it's better to keep it in the platform port.

File: plat/arm/common/aarch64/arm_common.c (hunk 1 above)
Comment (sorenb-xlnx) [inline plat/arm/common/aarch64/arm_common.c]: We compile the file plat/arm/common/aarch64/arm_common.c as we use some other content from it. As the symbols like *ARM_SYS_TIMCTL_BASE* are not defined on ZynqMP, compiling that file fails due to undefined identifiers - if not removed before compilation.

File: drivers/cadence/uart/cdns_console.S (hunk 4 above)
Comment (sorenb-xlnx) [inline drivers/cadence/uart/cdns_console.S]: There is some *rXpY* revision information (see: https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/serial/cdns,uart.txt), which I'm not sure whether it should go into the filename as the differences might be small and not requiring a whole new driver for each rev.
I'm not aware of any other SoC using this IP. I can move it to plat/xilinx/drivers or something like that.

File: plat/arm/common/aarch64/arm_common.c (hunk 1 above)
Comment (danh-arm) [inline plat/arm/common/aarch64/arm_common.c]: Ah yes, I remember now we discussed this before.

It might be slightly safer to wrap the whole function, not just the scoped contents. That way, a link error will be raised if a platform tries to use `arm_configure_sys_timer()` without defining `ARM_SYS_TIMCTL_BASE`.

File: include/drivers/cadence/cdns_uart.h (hunk 3 above)
Comment (sorenb-xlnx) [inline include/drivers/cadence/cdns_uart.h]: On Wed, 2016-03-30 at 08:23:27 -0700, danh-arm wrote:
> > @@ -0,0 +1,74 @@
> > +/*
> > + * Copyright (c) 2014, Xilinx Inc.
> > + * Written by Edgar E. Iglesias.
> 
> Please use the standard boilerplate text

Yep, this one slipped through. I'll fix this.

	Sören

File: plat/xilinx/zynqmp/aarch64/zynqmp_common.c [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+	/* Global timer init - Program time stamp reference clk */
+	val = mmio_read_32(CRL_APB_TIMESTAMP_REF_CTRL);
+	val |= CRL_APB_TIMESTAMP_REF_CTRL_CLKACT_BIT;
+	mmio_write_32(CRL_APB_TIMESTAMP_REF_CTRL, val);
+
+	/* Program freq register in System counter  and enable system counter. */
+	mmio_write_32(IOU_SCNTRS_BASEFREQ, zynqmp_get_system_timer_freq());
+	mmio_write_32(IOU_SCNTRS_CONTROL, IOU_SCNTRS_CONTROL_EN);
+
+	arm_config.flags |= ARM_CONFIG_HAS_INTERCONNECT;
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/aarch64/zynqmp_common.c]: Since the config is always the same, do you need to use arm_config at all?

File: plat/xilinx/zynqmp/aarch64/zynqmp_helpers.S [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	str	w0, [x1, #GICC_CTLR]
+
+	/*
+	 * There is no sane reason to come out of this wfi. This
+	 * cpu will be powered on and reset by the cpu_on pm api
+	 */
+	dsb	sy
+1:
+	wfi
+	b	1b
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/aarch64/zynqmp_helpers.S]: minor: fvp port (that this is derived from) now calls plat_panic_handler() here.

File: plat/xilinx/zynqmp/include/plat_macros.S [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+#include "../zynqmp_def.h"
+
+	/* ---------------------------------------------
+	 * The below required platform porting macro
+	 * prints out relevant GIC registers whenever an
+	 * unhandled exception is taken in BL31.
+	 * Clobbers: x0 - x10, x16, sp
+	 * ---------------------------------------------
+	 */
+	.macro plat_print_gic_regs
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/include/plat_macros.S]: You've got a conflict coming here with this https://github.com/ARM-software/arm-trusted-firmware/pull/549 (depending on which gets merged first).

File: plat/xilinx/zynqmp/plat_psci.c [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+
+void zynqmp_cpu_standby(plat_local_state_t cpu_state)
+{
+	VERBOSE("%s: cpu_state: 0x%x\n", __func__, cpu_state);
+
+	dsb();
+	wfi();
+}
+
+static int32_t zynqmp_nopmu_pwr_domain_on(u_register_t mpidr)
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/plat_psci.c]: minor: it would be better if these declarations exactly matched the plat_psci_ops (e.g. return int, not int32_t). Generally prefer to use the built in C types (e.g. int) where possible.

Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/plat_psci.c]: On Wed, 2016-03-30 at 08:54:32 -0700, danh-arm wrote:
> > +#include <psci.h>
> > +#include "pm_api_sys.h"
> > +#include "pm_client.h"
> > +#include "zynqmp_private.h"
> > +
> > +uintptr_t zynqmp_sec_entry;
> > +
> > +void zynqmp_cpu_standby(plat_local_state_t cpu_state)
> > +{
> > +	VERBOSE("%s: cpu_state: 0x%x\n", __func__, cpu_state);
> > +
> > +	dsb();
> > +	wfi();
> > +}
> > +
> > +static int32_t zynqmp_nopmu_pwr_domain_on(u_register_t mpidr)
> 
> minor: it would be better if these declarations exactly matched the plat_psci_ops (e.g. return int, not int32_t). Generally prefer to use the built in C types (e.g. int) where possible.

Agreed, and I wonder how it happened that they don't. Anyways, I'll
convert that to the appropriate types.
I'll keep the uint32_t for variables holding the return of
mmio_read_32() though, or should that also use 'unsigned int'?

	Thanks,
	Sören

File: plat/xilinx/zynqmp/aarch64/zynqmp_helpers.S (hunk 6 above)
Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/aarch64/zynqmp_helpers.S]: On Wed, 2016-03-30 at 08:44:12 -0700, danh-arm wrote:
> > +	mrs	x0, mpidr_el1
> > +
> > +	/* Deactivate the gic cpu interface */
> > +	ldr	x1, =BASE_GICC_BASE
> > +	mov	w0, #(IRQ_BYP_DIS_GRP1 | FIQ_BYP_DIS_GRP1)
> > +	orr	w0, w0, #(IRQ_BYP_DIS_GRP0 | FIQ_BYP_DIS_GRP0)
> > +	str	w0, [x1, #GICC_CTLR]
> > +
> > +	/*
> > +	 * There is no sane reason to come out of this wfi. This
> > +	 * cpu will be powered on and reset by the cpu_on pm api
> > +	 */
> > +	dsb	sy
> > +1:
> > +	wfi
> > +	b	1b
> 
> minor: fvp port (that this is derived from) now calls plat_panic_handler() here.

Thanks, for the heads up. I'll change it in our port as well.

	Sören

File: plat/arm/common/aarch64/arm_common.c (hunk 1 above)
Comment (sorenb-xlnx) [inline plat/arm/common/aarch64/arm_common.c]: On Wed, 2016-03-30 at 08:36:25 -0700, danh-arm wrote:
> > @@ -150,6 +151,7 @@ void arm_configure_sys_timer(void)
> >  
> >  	reg_val = (1 << CNTNSAR_NS_SHIFT(PLAT_ARM_NSTIMER_FRAME_ID));
> >  	mmio_write_32(ARM_SYS_TIMCTL_BASE + CNTNSAR, reg_val);
> > +#endif
> 
> Ah yes, I remember now we discussed this before.
> 
> It might be slightly safer to wrap the whole function, not just the scoped contents. That way, a link error will be raised if a platform tries to use `arm_configure_sys_timer()` without defining `ARM_SYS_TIMCTL_BASE`.

I'll give that a try. Though, if this is referenced from some common code,
an empty dummy might be necessary to be provided. I'll let you know how
it goes.

	Sören

File: plat/xilinx/zynqmp/plat_psci.c [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+	while (1)
+		wfi();
+}
+
+int zynqmp_validate_power_state(unsigned int power_state,
+				psci_power_state_t *req_state)
+{
+	VERBOSE("%s: power_state: 0x%x\n", __func__, power_state);
+
+	// FIXME: populate req_state
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/plat_psci.c]: Perhaps could use arm_validate_power_state() here

File: plat/xilinx/zynqmp/aarch64/zynqmp_common.c (hunk 5 above)
Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/aarch64/zynqmp_common.c]: On Wed, 2016-03-30 at 08:38:16 -0700, danh-arm wrote:
[...]
> > +
> > +	arm_config.flags |= ARM_CONFIG_HAS_INTERCONNECT;
> 
> Since the config is always the same, do you need to use arm_config at all?

I guess not. I think this is due to c&p and without looking at the
details I assumed the arm_config struct was used in some common core
code. Apparently that is not the case, hence, I'll remove the arm_config
stuff from our port.

	Thanks,
	Sören

File: plat/xilinx/zynqmp/pm_service/pm_api_sys.c [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+ *
+ * This is a blocking call, it will return only once PMU has responded.
+ * On a wakeup, resume address will be automaticaly set by PMU.
+ *
+ * @return	Returns status, either success or error+reason
+ */
+enum pm_ret_status pm_self_suspend(const enum pm_node_id nid,
+				   const uint32_t latency,
+				   const uint8_t state,
+				   const uint64_t address)
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/pm_service/pm_api_sys.c]: minor. Prefer built-in C types (especially int) where possible. Also, what value is there in making these const? Similarly for other funcs in this file.

File: plat/xilinx/zynqmp/pm_service/pm_svc_main.c [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+		ret = pm_mmio_write(pm_arg[0], pm_arg[1], pm_arg[2]);
+		SMC_RET1(handle, (uint64_t)ret);
+
+	case PM_MMIO_READ:
+	{
+		uint32_t value;
+
+		ret = pm_mmio_read(pm_arg[0], &value);
+		SMC_RET1(handle, (uint64_t)ret | ((uint64_t)value) << 32);
+	}
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/pm_service/pm_svc_main.c]: You're providing unfettered normal world access to this "protected" PM peripheral with no input checking. Is that definitely what you want?

File: plat/xilinx/zynqmp/sip_svc_setup.c [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+{
+	/* Let PM SMC handler deal with PM-related requests */
+	if (is_pm_fid(smc_fid)) {
+		return pm_smc_handler(smc_fid, x1, x2, x3, x4, cookie, handle,
+				      flags);
+	}
+
+	switch (smc_fid) {
+	case ZYNQMP_SIP_SVC_CALL_COUNT:
+		SMC_RET1(handle, PM_API_MAX);
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/sip_svc_setup.c]: Shouldn't this be + 3 (for the 3 SMCs in this function).

Comment (danh-arm) [inline plat/xilinx/zynqmp/sip_svc_setup.c]: Actually maybe +2 since the pm_api_id enum starts from 1.

File: plat/xilinx/zynqmp/zynqmp_private.h [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+uint32_t zynqmp_is_pmu_up(void);
+
+/*
+ * Register handler to specific GIC entrance
+ * for INTR_TYPE_EL3 type of interrupt
+ */
+int request_intr_type_el3(uint32_t, interrupt_type_handler_t);
+
+/* Declaration of function used as wake-up entrypoint */
+void bl31_entrypoint(void);
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/zynqmp_private.h]: Why is this needed? Doesn't seem to be used.

File: plat/xilinx/zynqmp/zynqmp_private.h [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+/* ZynqMP specific functions */
+uint32_t zynqmp_get_uart_clk(void);
+uint32_t zynqmp_is_pmu_up(void);
+
+/*
+ * Register handler to specific GIC entrance
+ * for INTR_TYPE_EL3 type of interrupt
+ */
+int request_intr_type_el3(uint32_t, interrupt_type_handler_t);
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/zynqmp_private.h]: Doesn't seem to be used or defined.

File: plat/xilinx/zynqmp/aarch64/zynqmp_common.c [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+}
+
+/*
+ * A single boot loader stack is expected to work on both the Foundation ZYNQMP
+ * models and the two flavours of the Base ZYNQMP models (AEMv8 & Cortex). The
+ * SYS_ID register provides a mechanism for detecting the differences between
+ * these platforms. This information is stored in a per-BL array to allow the
+ * code to take the correct path.Per BL platform configuration.
+ */
+int zynqmp_config_setup(void)
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/aarch64/zynqmp_common.c]: Probably should return void

File: plat/xilinx/zynqmp/plat_psci.c (hunk 8 above)
Comment (danh-arm) [inline plat/xilinx/zynqmp/plat_psci.c]: Yes, it's OK to use the exact width types when you need them (e.g. storing known width register data, packing data in structs, etc...).

File: drivers/cadence/uart/cdns_console.S (hunk 4 above)
Comment (danh-arm) [inline drivers/cadence/uart/cdns_console.S]: Actually, it's fine - leave it here.

File: plat/xilinx/zynqmp/zynqmp_private.h (hunk 14 above)
Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/zynqmp_private.h]: On Wed, 2016-03-30 at 10:04:57 -0700, danh-arm wrote:
> > +
> > +int zynqmp_config_setup(void);
> > +
> > +void zynqmp_cci_init(void);
> > +void zynqmp_cci_enable(void);
> > +void zynqmp_cci_disable(void);
> > +
> > +/* ZynqMP specific functions */
> > +uint32_t zynqmp_get_uart_clk(void);
> > +uint32_t zynqmp_is_pmu_up(void);
> > +
> > +/*
> > + * Register handler to specific GIC entrance
> > + * for INTR_TYPE_EL3 type of interrupt
> > + */
> > +int request_intr_type_el3(uint32_t, interrupt_type_handler_t);
> 
> Doesn't seem to be used or defined.

Yeah, some stuff that will remain out-of-tree for a little longer and I
missed the declaration in the header, sorry. I'll remove that from the
upstream submission.

	Sören

File: plat/xilinx/zynqmp/zynqmp_private.h (hunk 13 above)
Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/zynqmp_private.h]: On Wed, 2016-03-30 at 10:03:59 -0700, danh-arm wrote:
> > +void zynqmp_cci_init(void);
> > +void zynqmp_cci_enable(void);
> > +void zynqmp_cci_disable(void);
> > +
> > +/* ZynqMP specific functions */
> > +uint32_t zynqmp_get_uart_clk(void);
> > +uint32_t zynqmp_is_pmu_up(void);
> > +
> > +/*
> > + * Register handler to specific GIC entrance
> > + * for INTR_TYPE_EL3 type of interrupt
> > + */
> > +int request_intr_type_el3(uint32_t, interrupt_type_handler_t);
> > +
> > +/* Declaration of function used as wake-up entrypoint */
> > +void bl31_entrypoint(void);
> 
> Why is this needed? Doesn't seem to be used.

I think that's a left over from our initial PSCI implementation and was
obsoleted with the migration to v1.2. I'll remove this.

	Thanks,
	Sören

File: plat/xilinx/zynqmp/pm_service/pm_svc_main.c (hunk 11 above)
Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/pm_service/pm_svc_main.c]: On Wed, 2016-03-30 at 09:54:24 -0700, danh-arm wrote:
> > +		SMC_RET1(handle, (uint64_t)ret |
> > +			 ((uint64_t)reset_status << 32));
> > +	}
> > +
> > +	/* PM memory access functions */
> > +	case PM_MMIO_WRITE:
> > +		ret = pm_mmio_write(pm_arg[0], pm_arg[1], pm_arg[2]);
> > +		SMC_RET1(handle, (uint64_t)ret);
> > +
> > +	case PM_MMIO_READ:
> > +	{
> > +		uint32_t value;
> > +
> > +		ret = pm_mmio_read(pm_arg[0], &value);
> > +		SMC_RET1(handle, (uint64_t)ret | ((uint64_t)value) << 32);
> > +	}
> 
> You're providing unfettered normal world access to this "protected" PM peripheral with no input checking. Is that definitely what you want?

All access checking is currently deferred to the underlying FW layer. So
far, we haven't seen a need to differentiate between S or NS requests
there, hence ATF doesn't need to check anything for now.

	Sören

File: plat/xilinx/zynqmp/aarch64/zynqmp_common.c (hunk 15 above)
Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/aarch64/zynqmp_common.c]: On Wed, 2016-03-30 at 10:05:41 -0700, danh-arm wrote:
> > + *
> > + * Return 0 if firmware is not available, non 0 otherwise
> > + */
> > +uint32_t zynqmp_is_pmu_up(void)
> > +{
> > +	return zynqmp_pmufw_present;
> > +}
> > +
> > +/*
> > + * A single boot loader stack is expected to work on both the Foundation ZYNQMP
> > + * models and the two flavours of the Base ZYNQMP models (AEMv8 & Cortex). The
> > + * SYS_ID register provides a mechanism for detecting the differences between
> > + * these platforms. This information is stored in a per-BL array to allow the
> > + * code to take the correct path.Per BL platform configuration.
> > + */
> > +int zynqmp_config_setup(void)
> 
> Probably should return void

OK

File: plat/xilinx/zynqmp/include/plat_macros.S (hunk 7 above)
Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/include/plat_macros.S]: On Wed, 2016-03-30 at 08:49:21 -0700, danh-arm wrote:
> > + */
> > +#ifndef __PLAT_MACROS_S__
> > +#define __PLAT_MACROS_S__
> > +
> > +#include <arm_macros.S>
> > +#include <cci_macros.S>
> > +#include "../zynqmp_def.h"
> > +
> > +	/* ---------------------------------------------
> > +	 * The below required platform porting macro
> > +	 * prints out relevant GIC registers whenever an
> > +	 * unhandled exception is taken in BL31.
> > +	 * Clobbers: x0 - x10, x16, sp
> > +	 * ---------------------------------------------
> > +	 */
> > +	.macro plat_print_gic_regs
> 
> You've got a conflict coming here with this https://github.com/ARM-software/arm-trusted-firmware/pull/549 (depending on which gets merged first).

Thanks for the heads up. Seems there is not much that can be done to
prevent the conflict. We'll see how it goes.

File: plat/xilinx/zynqmp/pm_service/pm_api_sys.c (hunk 10 above)
Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/pm_service/pm_api_sys.c]: I'll clean up the file.

File: plat/xilinx/zynqmp/sip_svc_setup.c (hunk 12 above)
Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/sip_svc_setup.c]: Yep. I'll add 2.

File: plat/xilinx/zynqmp/plat_psci.c (hunk 9 above)
Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/plat_psci.c]: On Wed, 2016-03-30 at 09:24:13 -0700, danh-arm wrote:
[...]
> > +int zynqmp_validate_power_state(unsigned int power_state,
> > +				psci_power_state_t *req_state)
> > +{
> > +	VERBOSE("%s: power_state: 0x%x\n", __func__, power_state);
> > +
> > +	// FIXME: populate req_state
> 
> Perhaps could use arm_validate_power_state() here

I'll look into this.

File: plat/arm/common/aarch64/arm_common.c (hunk 1 above)
Comment (sorenb-xlnx) [inline plat/arm/common/aarch64/arm_common.c]: That seems to work. I'll wrap the whole function in #ifdefs then.

File: plat/xilinx/zynqmp/plat_psci.c (hunk 9 above)
Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/plat_psci.c]: On Wed, 2016-03-30 at 09:24:13 -0700, danh-arm wrote:
[...]
> > +int zynqmp_validate_power_state(unsigned int power_state,
> > +				psci_power_state_t *req_state)
> > +{
> > +	VERBOSE("%s: power_state: 0x%x\n", __func__, power_state);
> > +
> > +	// FIXME: populate req_state
> 
> Perhaps could use arm_validate_power_state() here

Using that function directly is hardly possible, as that pulls in a lot
things that are specific to the ARM platforms, duplicated functions,
etc.
Currently, I tend to defer implementing this function for ZynqMP.

	Sören

**Comment by sorenb-xlnx** [comment]
I updated the branch. Most comments so far have been addressed. And a few fixes/updates from our fork have been integrated into the PR (e.g. building with LOG_LEVEL=0 was broken).

File: plat/xilinx/zynqmp/aarch64/zynqmp_common.c [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+	case ZYNQMP_CSU_VERSION_EP108:
+		return 4000000;
+	case ZYNQMP_CSU_VERSION_QEMU:
+		return 50000000;
+	}
+
+	return 100000000;
+}
+
+#if LOG_LEVEL >= LOG_LEVEL_NOTICE
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/aarch64/zynqmp_common.c]: By default, NOTICE messages are printed out in release builds as well as debug builds. Here you seem to be outputting quite a log of platform information on the console during release builds. Could this information be useful to an attacker? Consider reducing the log level to INFO.

It's up to you whether you want to fix this or not (just let me know).

File: plat/xilinx/zynqmp/aarch64/zynqmp_common.c [hunk 17]
```diff
@@ -0,0 +1,324 @@
...
+{
+	uint64_t counter_base_frequency;
+
+	/* FIXME: Read the frequency from Frequency modes table */
+	counter_base_frequency = zynqmp_get_system_timer_freq();
+
+	return counter_base_frequency;
+}
+
+void zynqmp_cci_init(void)
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/aarch64/zynqmp_common.c]: minor: since you don't provide any special functionality here, you could call the `plat_arm_interconnect_xxx()` methods directly

**Comment by danh-arm** [comment]
It's not mandatory to fix my comments before merging. Just let me know if you want to rework or for me to merge as-is.

File: plat/xilinx/zynqmp/aarch64/zynqmp_common.c (hunk 16 above)
Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/aarch64/zynqmp_common.c]: Thanks for the suggestion. My assumption were that production builds
would like to set LOG_LEVEL=0 and that attackers would not have access
to a debug UART. I'll take this topic for further internal discussions.
I'd like to keep it the way it is for the PR though.

File: plat/xilinx/zynqmp/aarch64/zynqmp_common.c (hunk 17 above)
Comment (sorenb-xlnx) [inline plat/xilinx/zynqmp/aarch64/zynqmp_common.c]: Yes, I agree. There is no need for this additional layer currently. I
remove these functions and fix up all callsites to directly use
plat_arm_interconnect_foo.

**Comment by sorenb-xlnx** [comment]
Hi Dan,

I'll rework the CCI/interconnect stuff. It should be possible for me to
update the PR later today.

	Thanks,
	Sören

**Comment by sorenb-xlnx** [comment]
I made the discussed change regarding the CCI/interconnect functions and updated the PR.

**Comment by danh-arm** [comment]
jenkins: test this please

File: plat/xilinx/zynqmp/pm_service/pm_ipi.c [hunk 18]
```diff
@@ -0,0 +1,217 @@
+#include <bakery_lock.h>
```
Comment (danh-arm) [inline plat/xilinx/zynqmp/pm_service/pm_ipi.c]: This file contains no copyright/license boilerplate text

**Comment by danh-arm** [comment]
Our CI loop flagged checkpatch errros (which must be fixed) and warnings (which may be fixed):

```
ERROR: trailing whitespace
#54: FILE: docs/plat/xilinx-zynqmp.md:9:
+BL31 is ATF.  $

ERROR: trailing whitespace
#55: FILE: docs/plat/xilinx-zynqmp.md:10:
+BL32 is an optional Secure Payload.  $

ERROR: trailing whitespace
#56: FILE: docs/plat/xilinx-zynqmp.md:11:
+BL33 is the non-secure world software (U-Boot, Linux etc).  $
...
ERROR: code indent should use tabs where possible
#294: FILE: plat/xilinx/zynqmp/aarch64/zynqmp_common.c:194:
+^I            ZYNQMP_CSU_IDCODE_FAMILY << ZYNQMP_CSU_IDCODE_FAMILY_SHIFT |$
...
ERROR: code indent should use tabs where possible
#295: FILE: plat/xilinx/zynqmp/aarch64/zynqmp_common.c:195:
+^I            ZYNQMP_CSU_IDCODE_REVISION << ZYNQMP_CSU_IDCODE_REVISION_SHIFT;$
...
ERROR: do not use C99 // comments
#1170: FILE: plat/xilinx/zynqmp/plat_psci.c:311:
+	// FIXME: populate req_state

ERROR: do not use C99 // comments
#1178: FILE: plat/xilinx/zynqmp/plat_psci.c:319:
+	// FIXME: Actually validate
...
total: 7 errors, 81 warnings, 3722 lines checked
```

**Comment by danh-arm** [comment]
Sorry Sören - our CI loops picked up some more issues that need fixing. I should have run this earlier.

Dan.

**Comment by sorenb-xlnx** [comment]
On Wed, 2016-04-06 at 04:43:11 -0700, danh-arm wrote:
> Sorry Sören - our CI loops picked up some more issues that need fixing. I should have run this earlier.

No big deal. I should be able to get that done today. Looks mostly like
trivial formatting issues. Though these:
  ERROR: trailing whitespace
  #54: FILE: docs/plat/xilinx-zynqmp.md:9:
  +BL31 is ATF.  $
  
  ERROR: trailing whitespace
  #55: FILE: docs/plat/xilinx-zynqmp.md:10:
  +BL32 is an optional Secure Payload.  $
  
  ERROR: trailing whitespace
  #56: FILE: docs/plat/xilinx-zynqmp.md:11:
  +BL33 is the non-secure world software (U-Boot, Linux etc).  $

are markdown formatting. These will remain. I'm not sure if these test
are gating, but it might be a good idea to relax the rules for markdown
text.

	Sören

**Comment by danh-arm** [comment]
> These will remain.

But it's trivial to remove those trailing spaces.

checkpatch errors are currently gating. I agree this is a bit harsh for markdown documents but it's going to require some effort to apply a different policy to some files compared to the rest of the code, and I'd rather not exclude these files from the check altogether.

**Comment by sorenb-xlnx** [comment]
On Wed, 2016-04-06 at 08:46:31 -0700, danh-arm wrote:
> > These will remain.
> 
> But it's trivial to remove those trailing spaces.

Is it? Would you tell me how? The trailing whitespaces are
causing a line break. I'm not aware of another method for
forcing a line break in github markdown.

	Thanks,
	Sören

**Comment by danh-arm** [comment]
>  Would you tell me how?
Sure, I'm just talking about the 2 trailing space (0x20) chars at the end of the 3 lines in question, not the line break + carriage return chars. Here's a fixed version (which I had to save as txt 'cos GitHub doesn't let me attach md files bizarrely).

[xilinx-zyncmp.txt](https://github.com/ARM-software/arm-trusted-firmware/files/206919/xilinx-zyncmp.txt)

**Comment by danh-arm** [comment]
Sorry, I misunderstood you. You're talking about the spaces forcing a line break in the _rendered _file. I didn't know about that trick; I see that this is official markdown syntax so yes, you shouldn't have to fix this!

I'll let this go for now and think about how to improve our CI loop later.

**Comment by sorenb-xlnx** [comment]
I updated the PR. That should fix the missing license header, checkpatch errors - apart from the discussed markdown ones - and some checkpatch warnings.

**Comment by danh-arm** [comment]
jenkins:retest this please

**Comment by danh-arm** [comment]
jenkins:retest this please
